### PR TITLE
Allow for date fields to have a date_format in facets.

### PR DESCRIPTION
**Pull request by jordandukart**
**JIRA Ticket**: https://jira.duraspace.org/browse/ISLANDORA-1840

# What does this Pull Request do?
When displaying a date field that it's a facet having a human readable label is more user friendly as opposed to a ISO8601 datestamp.

# What's new?
Expose the format_date for all date fields in facets and don't constrain it just to date range facets. .

# How should this be tested?
* Create a facet field from a date
* Ensure that it's not a range field
* Enter a date format such as `Y-m-d`
* Search for some content, note how the formatted date is displayed as opposed to the full timestamp 

# Additional Notes:
Bubbling up the date formatter won't have any impact in the stored settings as the form isn't `#tree`ed.

# Interested parties:
@Islandora/7-x-1-x-committers as I am maintainer.

### CODE REVIEW

File: includes/admin.inc
Comment (qadan) [inline includes/admin.inc]: Didja mean to link [PHP date formatting](http://php.net/manual/en/datetime.formats.php) docs instead of Lucene DateMathParser docs?

Comment (whikloj) [reply includes/admin.inc]: That's actually been in there for years based on the copied block below. But a good catch nonetheless.

**Comment by willtp87** [comment]
This is looking good to me.

**Comment by jordandukart** [comment]
Bump @Islandora/7-x-1-x-committers

**Comment by DiegoPino** [comment]
@jordandukart will check this later tonight, sorry can not make it sooner.

**Comment by jordandukart** [comment]
Bump, @Islandora/7-x-1-x-committers

**Comment by whikloj** [comment]
So this works as expected, however I have a question about the desired outcome.

I ingested a bunch of items with ten-million-with-a-hat. So they all have similar but slightly different created dates. When I use this I can adjust the date to appear with the desired format, but as it doesn't impact the underlying facet the outcome is a little unexpected.

For example without the PR I get create date facets like this:
facet_before

After I can set it to use **Y-m-d** which ends up with.
facet_adjusted

But even though the facets all _look_ the same they have hidden values which a user can't know until they click on it.
facet_result

I'm not sure what the feature request was for this, but I'm wondering if this will confuse things?

**Comment by DiegoPino** [comment]
@jordandukart @whikloj  itis a bit confusing yes. Maybe the logic should group on the resulting facet after the format and not on the original values with format as label?

**Comment by jordandukart** [comment]
I feel like this is dependent on the source data and configuration. For example if you had a field with

```
<dateIssued>1865</dateIssued>
```

after the JODA conversions this will get indexed as 1865-01-01T00:00:00Z, a Y format here makes sense.

In @whikloj's example I feel like the date format defined is lossey and the resulting UX is convulated and weird (as noted). Perhaps this can be clarified within the `#description` of the configuration for the field? 

If I am interpreting @DiegoPino's suggestion it would seem that grouping facets is better accomplished with the date range functionality that exists within Solr and not post-processing in PHP.

**Comment by whikloj** [comment]
@jordandukart yeah I realize my conversion is not ideal. Trying to think how someone might use it. Perhaps ensuring there is a clear `#description` that you should only use this to format the date using all non-empty parts (or some better words) is the answer. 

@DiegoPino I don't know that I like the idea of re-grouping them together as that would be done on the PHP side and slow things down.  We could certainly do that, but I think people should be spending more time ensuring they have good and useful Solr fields.

**Comment by jordandukart** [comment]
@whikloj: Ya I agree, it's super dependent on source data but exposing the configuration to let an admin user rule over it seems to fall in the realm of knowing what they are configuring. Can update the `#description` accordingly.

**Comment by DiegoPino** [comment]
@whikloj ok, i understand your concern but this pull affects/alters via the "php side of things", it's not a solr formating via solr. It's solr data parsed via php functions. Maybe putting a hook somewhere would be a more generic approeach and then any dev/islandora shop - via external modules - could just affect the output of facets. I feel since this is very dependant on the data itself, is not really a change that can be easily generalised. But i leave this to your jugment, i'm ok if at least the wording changes as @rosiel proposes.

**Comment by whikloj** [comment]
Those are fair points @DiegoPino. I'd really rather not open up that can of worms unless we have more use cases for this type of facet altering.

@jordandukart considering @DiegoPino's suggestion and my concerns around it and around confusion using this feature, along with the use case you mentioned and the fact we both agree that this really depends on the source data. Have you considered indexing the dateIssued part as a string as well as a date and then using the appropriate option to facet on. ie. use date field for a range facet or string field for a simple part of a date?

**Comment by jordandukart** [comment]
Updated the description.

@DiegoPino: I agree that a hook to alter portions of the Solr query processor class in PHP would be beneficial but would be best suited for discussion and implementation in another ticket.

@whikloj: That's an acceptable approachable for sure, didn't cross my mind at the time of writing the XSL for indexing. With that in mind I feel like exposing this optional configuration is still beneficial to the module as a whole.

**Comment by whikloj** [comment]
@jordandukart Should this date format also apply to the filters in the breadcrumbs?

**Comment by whikloj** [comment]
Like once someone clicks on a facet, the breadcrumb shows the original date format. Should it continue to display the formatted date?

**Comment by jordandukart** [comment]
Updated the breadcrumb label.

**Comment by whikloj** [comment]
Looks good, waiting on Travis.

**Review by DiegoPino — APPROVED** [review]
Looks good.

**Pull request merged**
